File: src/test/java/com/wafflestudio/draft/api/RegionApiControllerTest.java
package com.wafflestudio.draft.api;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.http.MediaType;
import org.springframework.security.test.context.support.WithMockUser;
import org.springframework.test.context.junit.jupiter.SpringExtension;
import org.springframework.test.web.servlet.MockMvc;

import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.collection.IsCollectionWithSize.hasSize;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;


// FIXME: This test is depending on DataLoader and other logics, which is not desirable as genuine unit test.
// @ActiveProfiles("test")
@ExtendWith(SpringExtension.class)
@SpringBootTest
@AutoConfigureMockMvc
public class RegionApiControllerTest {

    @Autowired
    private MockMvc mockMvc;

    @Test
    @WithMockUser
    public void getRegionRoomTest() throws Exception {
        this.mockMvc.perform(get("/api/v1/region/room/").contentType(MediaType.APPLICATION_JSON))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$", hasSize(1)))
                .andExpect(jsonPath("$[0].id", is(1)))
                .andExpect(jsonPath("$[0].name", is("TEST_REGION")))
                .andExpect(jsonPath("$[0]['rooms']", hasSize(5)))
                .andExpect(jsonPath("$[0]['rooms'][4].name", is("TEST_ROOM_5")))
                .andExpect(jsonPath("$[0]['rooms'][4]['participants']", hasSize(1)));
    }
}

File: src/test/java/com/wafflestudio/draft/api/RegionApiControllerTest.kt
package com.wafflestudio.draft.api

import org.hamcrest.CoreMatchers
import org.hamcrest.collection.IsCollectionWithSize
import org.junit.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.http.MediaType
import org.springframework.security.test.context.support.WithMockUser
import org.springframework.test.context.junit.jupiter.SpringExtension
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders
import org.springframework.test.web.servlet.result.MockMvcResultMatchers

// FIXME: This test is depending on DataLoader and other logics, which is not desirable as genuine unit test.
// @ActiveProfiles("test")
@ExtendWith(SpringExtension::class)
@SpringBootTest
@AutoConfigureMockMvc
class RegionApiControllerTest(
        private val mockMvc: MockMvc
) {

    @get:Throws(Exception::class)
    @get:WithMockUser
    @get:Test
    val regionRoomTest: Unit
        get() {
            mockMvc.perform(MockMvcRequestBuilders.get("/api/v1/region/room/").contentType(MediaType.APPLICATION_JSON))
                    .andExpect(MockMvcResultMatchers.status().isOk)
                    .andExpect(MockMvcResultMatchers.jsonPath("$['results']", IsCollectionWithSize.hasSize<Any>(1)))
                    .andExpect(MockMvcResultMatchers.jsonPath("$['results'][0].id", CoreMatchers.`is`(1)))
                    .andExpect(MockMvcResultMatchers.jsonPath("$['results'][0].name", CoreMatchers.`is`("TEST_REGION")))
                    .andExpect(MockMvcResultMatchers.jsonPath("$['results'][0]['rooms']", IsCollectionWithSize.hasSize<Any>(15)))
                    .andExpect(MockMvcResultMatchers.jsonPath("$['results'][0]['rooms'][4].name", CoreMatchers.`is`("TEST_ROOM_5")))
                    .andExpect(MockMvcResultMatchers.jsonPath("$['results'][0]['rooms'][4]['participants']", IsCollectionWithSize.hasSize<Any>(1)))
        }
}